packages/mcp/src/auth/implementations/oauth2-authorization-code.ts

```diff
@@ -0,0 +1,394 @@
+      OAuth2AuthCodeProvider.stateToProvider.set(state, this);
+
+      // Build authorization URL with PKCE
+      const authUrl = new URL(this.config.authorizationEndpoint);
+      authUrl.searchParams.set('response_type', 'code');
+      authUrl.searchParams.set('client_id', this.config.clientId);
+      authUrl.searchParams.set('redirect_uri', this.config.redirectUri);
+      authUrl.searchParams.set('state', state);
+      authUrl.searchParams.set('code_challenge', codeChallenge);
+      authUrl.searchParams.set('code_challenge_method', 'S256');
+
+      if (this.config.scope) {
+        authUrl.searchParams.set('scope', this.config.scope);
+      }
+
+      if (this.config.audience) {
+        authUrl.searchParams.set('audience', this.config.audience);
+      }
+
+      logEvent('info', 'auth:oauth_flow_initiated', {
+        authUrl: this.config.authorizationEndpoint,
+        redirectUri: this.config.redirectUri,
+        scope: this.config.scope,
+      });
+
+      // Log URL for user to open (following protocol - NO browser launch package)
+      console.info('\n🔐 Please open this URL in your browser to authorize:');
+      console.info(authUrl.toString());
+      console.info('\nWaiting for authorization callback...');
+      console.info(`Timeout in ${this.AUTH_TIMEOUT_MS / 1000} seconds\n`);
+
+      // Also log structured event for monitoring
+      logEvent('info', 'auth:oauth_authorization_required', {
+        authUrl: authUrl.toString(),
+        timeout: this.AUTH_TIMEOUT_MS / 1000,
+      });
+    });
+  }
+
+  /**
+   * Exchange authorization code for access token
+   */
+  private async exchangeCodeForTokenResponse(
+    code: string,
+    codeVerifier: string,
+  ): Promise<OAuth2TokenResponse> {
+    const body = new URLSearchParams({
+      grant_type: 'authorization_code',
+      code,
+      redirect_uri: this.config.redirectUri,
+      client_id: this.config.clientId,
+      code_verifier: codeVerifier,
+    });
+
+    const headers: Record<string, string> = {
+      'Content-Type': 'application/x-www-form-urlencoded',
+    };
+
+    // Add client authentication if client secret is provided
+    if (this.config.clientSecret) {
+      const credentials = Buffer.from(
+        `${this.config.clientId}:${this.config.clientSecret}`,
+      ).toString('base64');
+      headers.Authorization = `Basic ${credentials}`;
+    }
+
+    try {
+      const response = await fetch(this.config.tokenEndpoint, {
+        method: 'POST',
+        headers,
+        body: body.toString(),
+      });
+
+      if (!response.ok) {
+        await this.handleTokenRequestError(undefined, response);
+      }
+
+      const tokenResponse = (await response.json()) as OAuth2TokenResponse;
+
+      this.validateTokenResponse(tokenResponse);
+
+      return tokenResponse;
+    } catch (error) {
+      await this.handleTokenRequestError(error);
+      throw new Error('This line should never be reached');
```

Comment:
Code smell: Same issue as in oauth2-client-credentials.ts. This line should never be reached but it's after a call to `handleTokenRequestError` which should throw. Consider removing this unreachable code.